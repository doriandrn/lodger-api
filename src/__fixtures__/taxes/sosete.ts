import { FieldsCreator } from "~/lib/Field";

type Sosete = {
  numar: Number,
  lungime: Number
}

const fields: FieldsCreator<Sosete> = {
  numar: {
    type: 'number',
    index: true
  },
  lungime: {
    type: 'number'
  }
}

export default {
  fields
}
