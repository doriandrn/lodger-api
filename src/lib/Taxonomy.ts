import { Taxonomii, notify } from "index";
import { RxDocument, RxCollection, RxDatabase, RxDocumentBase, RxQuery } from 'rxdb'
import { Form } from './Form'
import Subscriber from './Subscriber'
import LodgerConfig from 'lodger.config'
import TaxonomyError from './Error'

import { predefinite } from 'forms/serviciu'
import { RootState } from "./Store";
import { GetterTree, ActionTree, Dispatch, Store } from 'vuex'


 /**
  * Taxonomy item
  *
  * @interface LodgerTaxonomy
  */
 interface LodgerTaxonomy<N extends Taxonomie, S> {
  readonly plural: Plural<N>,
  readonly hasReference: Boolean,
  // readonly referenceTaxonomies: Taxonomy<Taxonomie>[],
  // readonly dependantTaxonomies: Taxonomy<Taxonomie>[],

  readonly collection: RxCollection<N>,
  readonly store: Store<S>

  put (data: Object): Promise<RxDocument<N>> | void
  trash (id: string): Promise<RxDocument<N>>
  select (id: string, subscriberName: string): void
}


/**
 *
 *
 * @interface SubscribableTaxonomy
 * @extends {LodgerTaxonomy<any, S>}
 * @template N
 * @template S
 */
interface SubscribableTaxonomy<N, S> extends LodgerTaxonomy<any, S> {
  readonly subscribers: SubscriberList<N>
  readonly data: SubscriberDataHolder
  readonly subscribed: boolean,

  subscribe (name: string, criteriu ?: Criteriu): Promise<Subscriber<N>>
  unsubscribeAll: (subscriberName?: string) => void

  onFirstTimeInit: () => void
}


interface SearchableTaxonomy extends SubscribableTaxonomy {
  searchMap: Map<string, string>
  searchResults: SearchResults

  search (input: string): Promise<SearchResults>
}

type SubscriberList<N> = {
  [k: string]: Subscriber<N>[]
}

type SearchResults = {
  [k: string]: Result[]
}

// a search result
type Result = {
  id: string,
  value: string,
  relevance: number
}

/**
 * @class Taxonomy
 * @implements {LodgerTaxonomy}
 *
 * @requires Form
 *
 * @param {Taxonomie} name - name of the form
 * @param {Form} form - the constructed form item
 */
class Taxonomy<T extends Taxonomie> implements LodgerTaxonomy<T> {

  /**
   * Creates an instance of Taxonomy.
   *
   * @param {Taxonomie} name
   * @param {Form} form
   * @memberof Taxonomy
   */
  constructor (
    protected collection: RxCollection<T>,
    protected store: Store<T>
  ) {
  }

  /**
   * Removes a Document by ID from the collection
   *
   * @param {string} id
   * @returns {RxDocument<T>} removed document
   * @memberof Taxonomy
   */
  async trash (id: string) {
    const doc: RxQuery<T, RxDocument<T>> = await this.collection.findOne(id)
    return await doc.remove()
  }

  /**
   * Inserts/upserts a new item in DB
   *
   * @param {Object} data
   * @returns {RxDocument<Taxonomie>} the fresh document
   *
   * @memberof Taxonomy
   */
  async put (
    data: { _id ?: string},
  ) {
    if (!data || Object.keys(data).length < 1)
      throw new TaxonomyError(Errors.missingData, data)

    const { collection, form } = this

    /**
     * If form submitted with an _id, must be an upsert
     */
    const method = data._id ?
      'upsert' :
      'insert'

    /**
     * add references, default values, etc
     */
    const finalData = form.value(true)

    // const internallyHandledData = handleOnSubmit(data, { referencesIds, store })

    /**
     * do the insert / upsert and following actions
     */
    try {
      const doc = await collection[method](finalData)
      const id = doc._id
      this.store.dispatch(`set_last`, id)
      this.select({ doc, id, subscriber })

      notify({
        type: 'success',
        text: `pus ${taxonomy} ${id}`
      })
      return doc
    } catch (e) {
      notify({
        type: 'error', text: String(e)
      })
    }
  }

  /**
   * select an item
   * brings in the active Document from DB
   *
   * @param taxonomie
   * @param id
   */
  async select (
    data: SelectedItemData
  ) {
    const { plural, store: { dispatch } } = this
    const isObj = typeof data === 'object' && data !== null

    const id = isObj && data.id ? data.id : data
    const subscriber = isObj && data.subscriber ? data.subscriber : undefined

    await dispatch(`${taxonomie}/select`, id)

    // // deselect
    // if (!id) {
    //   await dispatch(`${taxonomie}/select`, undefined)
    //   return
    // }

    // // delay this, await for changes from rxdbb
    // const doc = isObj && data.doc ?
    //   data.doc :
    //   await vueHelper.getItem(plural, id, subscriber)

    // debug('selected doc', doc._id)

    // if (!doc) {
    //   throw new LodgerError('invalid id supplied on select %%', id)
    // } else {
    //   this._activeDocument = { taxonomie, doc }
    //   await dispatch(`${taxonomie}/select`, id)
    // }

    // on deselect, unsubscribe
    // if (id === null) await this.unsubscribe(plural, subscriber) //todo: use data.subscribe .unsubscribe()
  }

  // /**
  //  * Binds a RXCollection to taxonomy
  //  *
  //  * @memberof Taxonomy
  //  * @param {RxCollection<Taxonomie>} collection
  //  * @returns {RxCollection | undefined}
  //  */
  // get collection () {
  //   return this.collection
  // }

  // set collection (col: RxCollection<Taxonomie>) {
  //   this.collection = col
  // }

  /**
   * Reference taxonomies of a taxonomy
   * @memberof Taxonomy
   * @returns {Array} taxonomii
   */
  get referenceTaxonomies () {
    const { fields } = this.form

    return <Taxonomie[]>fields
      .filter(field => field.id.indexOf('Id') === field.id.length - 2)
      .map(field => field.id.replace('Id', ''))
  }

  /**
   * Checks for a reference taxonomy of taxonomy
   *
   * @readonly
   * @memberof Taxonomy
   * @returns {Boolean} has a reference taxonomy or not
   */
  get hasReference () {
    return true
  }



  /**
   * @readonly
   * @memberof Taxonomy
   * @returns {String} plural of taxonomy
   */
  get plural () {
    return this.form.plural
  }

  /**
   * @readonly
   * @memberof Taxonomy
   * Taxonomy default config
   */
  get config () {
    const { taxonomii } = LodgerConfig
    const { defaults } = taxonomii
    return taxonomii[this.plural] || defaults
  }

  get defaultCriteriu () {
    return this.config.criteriu
  }

  /**
   *
   *
   * @readonly
   * @memberof Taxonomy
   * @returns {Boolean} if taxonomy represents a multiple select choice
   */
  get isMultipleSelect () {
    return ['Serviciu', 'Contor'].indexOf(this.name) > -1
  }


}

class STaxonomy extends Taxonomy implements SubscribableTaxonomy {
  protected subscribers: SubscriberList

  constructor () {

  }


  /**
   * Returns all data from subscribers
   *
   * @readonly
   * @memberof Taxonomy
   */
  get data () {
    return
  }

  /**
   * @readonly
   * @memberof Taxonomy
   * @returns {Boolean} if subscribed anywhere
   */
  get subscribed () {
    return this.subscribers.length > 0
  }
  /**
   * Subscribes
   *
   * @param {string} [subscriberName='main']
   * @param {Criteriu} [criteriuCerut]
   * @returns {Promise<Subscriber<T>>} the unwatcher for subscriber
   * @memberof Taxonomy
   */
  subscribe (
    subscriberName : string = 'main',
    criteriuCerut ?: Criteriu
  ): Promise<Subscriber<T>> {
    return new Subscriber(subscriberName, this).subscribe(criteriuCerut)
  }

  /**
   * Kills all active listeners for a given subscriber name
   *
   * @param {string} [subscriberName='main']
   * @returns {Promise}
   * @memberof Taxonomy
   */
  unsubscribeAll (subscriberName: string = 'main') {
    const subscribers = this.actives.subscribers[subscriberName]

    return Promise.all(
      Object.keys(subscribers).map(async subscriber => {
        await subscribers[subscriber].unsubscribe()
      })
    )
  }

  async onFirstTimeSubscribe () {
    const { name, collection, store } = this
    switch (name) {
      // insert predefined services
      case 'Serviciu':
        predefinite.forEach(async denumire => { await collection.insert({ denumire }) })
        break

      // insert admin user
      case 'Utilizator':
        const { _id } = await collection.insert({
          name: 'Administrator',
          rol: 'admin'
        })
        store.dispatch('utilizator/set_active', _id)
        break
    }

    subscribedTaxes.push(name)
  }
}


// /**
//  * Pt taxonomia ceruta
//  * ia formul
//  * si tot ce are nevoie de Id de altceva
//  * se populeaza
//  *
//  * @param {Object} { references, getters }
//  * @returns {Object} eg { asociatieId: 'XXXX' }
//  */
// function assignRefIdsFromStore (context: any) {
//   const { references, getters } = context
//   if (!(references && references.length)) return

//   const refsObj = {}

//   references.map((tax: Taxonomie) => {
//     refsObj[`${tax}Id`] = getters[`${tax}/selected`]
//   })

//   return refsObj
// }


// const xx: LodgerTaxonomyCreator<Taxonomie> = new Taxonomy('asociatie', 'form', 'collection')
/**
 * gets the sorting options for tax
 * @returns an object with each key used as a sorting option
 */
// get sortOptions () {
//   const { indexables, name } = this

//   if (!['serviciu', 'contor'].indexOf(name)) {
//     indexables.push('la')
//   }

//   // TODO: !!! ia din common methods
//   const sorts = {}
//   indexables.forEach(indexable => {
//     const label = `sort.${indexable === 'name' ? 'az' : indexable}`
//     Object.assign(sorts, { [indexable]: { label } })
//   })

//   // debug(`${name} => sortable fields`, sorts)

//   return sorts
// }
/**
 * Items to be display to user,
 * @returns {Object} the keys of the fields: their position
 *
 */
// get __displayItemKeys () {
//   const { fields } = this.data

//   return Object.assign({}, ...fields
//     .filter(field => field.showInList)
//     .map(field => ({ [field.id]: field.showInList }) )
//   )
// }


// // Filters the documents array for the one with the id
// export const getRxDocumentById = (docs: RxDocument<Taxonomie, any>[], id: string) => {
//   if (!docs.length)
//     throw new LodgerError('Empty docs provided: %%')
//   const doc = docs.filter(doc => doc._id === id)[0]
//   if (!(doc && isRxDocument(doc)))
//     throw new LodgerError('No document found %%', { id })
//   return doc
// }

//
  /**
   * Active document for taxonomy
  */
//  protected set _activeDocument (docHolder: ActiveDocumentHolder) {
//   let { taxonomie, doc } = docHolder
//   const debug = Debug('lodger:_activeDocument')
//   const gName = `${taxonomie}/activeDoc`
//   const { store } = this

//   if (!store.getters.hasOwnProperty(gName)) {
//     Object.defineProperty(store.getters, gName, {
//       configurable: false,
//       get () { return doc },
//       set (newDoc) { doc = newDoc }
//     })
//   } else {
//     store.getters[gName] = doc
//   }
// }
