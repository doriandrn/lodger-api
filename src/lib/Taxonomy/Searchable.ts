import SubscribableTaxonomy from './Subscribable'

interface SearchableTaxonomy extends SubscribableTaxonomy {
  search (input: string): Promise<SearchResults> | void
}

// a search result
type Result = {
  id: string,
  value: string,
  relevance: number
}

type SubscribableTaxonomyConfig = {
  maxResults: 6
}

export default class STaxonomy extends SubscribableTaxonomy implements SearchableTaxonomy {
  protected map: Map<string, string> = new Map()
  readonly _results: Result[] = []

  constructor (protected config: SubscribableTaxonomyConfig) {
    super(...arguments)
  }

  clearResults () {
    Object.keys(this.results).forEach(result => this.results[result] = [])
  }

  get results () {
    return this._results
      .sort((a, b) => Number(a.relevance) - Number(b.relevance))
      .reverse()
      .slice(0, this.config.maxResults)
  }

  /**
   * Cauta in searchMap
   * @param input - string de cautat
   * @alias Taxonomy.search
   */
  search (
    input: string
  ) {
    if (!input) return

    const iterator = this.map.entries()

    for (let [key, value] of iterator) {
      if (typeof value === 'function') continue
      const relevance = string_similarity(String(input), value)
      const id = String(key)
      this._results.push({ id, relevance, value })
    }
  }
}

/**
 * Helpers
 */
export const get_bigrams = function (string: string) {
  var i, j, ref, s, v;
  s = string.toLowerCase();
  v = new Array(s.length - 1);
  for (i = j = 0, ref = v.length; j <= ref; i = j += 1) {
    v[i] = s.slice(i, i + 2);
  }
  return v;
};

export const string_similarity = function (str1: string, str2: string) {
  var hit_count, j, k, len, len1, pairs1, pairs2, union, x, y;
  if (str1.length > 0 && str2.length > 0) {
    pairs1 = get_bigrams(str1);
    pairs2 = get_bigrams(str2);
    union = pairs1.length + pairs2.length;
    hit_count = 0;
    for (j = 0, len = pairs1.length; j < len; j++) {
      x = pairs1[j];
      for (k = 0, len1 = pairs2.length; k < len1; k++) {
        y = pairs2[k];
        if (x === y) {
          hit_count++;
        }
      }
    }
    if (hit_count > 0) {
      return (2.0 * hit_count) / union;
    }
  }
  return 0.0;
};
