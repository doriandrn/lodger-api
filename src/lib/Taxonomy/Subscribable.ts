import Taxonomy, { TaxonomyCreator } from './'
import SubscribableTaxonomyError from '~/lib/Error'
import Subscriber from 'rxcollection-subscriber'
import { computed, reaction, observable } from 'mobx'
import merge from 'deepmerge'

/**
 *
 *
 * @interface SubscribableTaxonomy
 * @extends {LodgerTaxonomy<any, S>}
 * @template N
 * @template S
 */
interface SubscribableTaxonomy<T extends Taxonomie> {
  readonly subscribers: SubscriberList<T>
  readonly data: SubscriberDataHolder,
  readonly subscribed: boolean
}

type SubscriberList<T extends Taxonomie> = {
  [k: string]: Subscriber<T>
}

export default class STaxonomy<T extends Taxonomie, I>
extends Taxonomy<T, I>
implements SubscribableTaxonomy<T> {
  readonly subscribers: SubscriberList<T> = {}
  @observable _refsIds: {[k in Taxonomii]: string} = {}

  constructor () {
    super(...arguments)
  }

  static init () {
    return super.init.apply(this, arguments)
  }

  /**
   * Returns all data from subscribers
   *
   * @readonly
   * @memberof Taxonomy
   */
  get data () {
    return this.subscribers
  }

  @computed get refsIds () {
    return this._refsIds
  }

  /**
   * @readonly
   * @memberof Taxonomy
   * @returns {Boolean} if subscribed anywhere
   */
  get subscribed () {
    return Object.keys(this.subscribers).length > 0
  }

  get defaultCriteria () {
    return super.config.criteriu
  }

  /**
   * Subscribes.
   *
   * @param {string} [subscriberName='main']
   * @param {Criteriu} [criteriuCerut]
   * @returns {Promise<Subscriber<T>>} the unwatcher for subscriber
   * @memberof Taxonomy
   */
  async subscribe (
    subscriberName : string = 'main',
    options
  ): void {
    const {
      hooks,
      subscribers,
      plural,
      $lodger: {
        state,
        modal,
        taxonomies
      }
    } = this

    if (subscribers[subscriberName])
      return
      // throw new LodgerError('Cannot subscribe - A subscriber with this name already exists!')

    const descriptor = `${subscriberName}-${taxonomies.indexOf(plural)}`

    const subState = state.subs[descriptor] ||
      Object.assign(state.subs, { [descriptor]: {} }) && state.subs[descriptor]

    const sub = this.subscribers[subscriberName] = new Subscriber(this.collection, merge(options, subState))

    if (this.parents && this.parents.length && !sub.refsIds) {
      sub.refsIds = observable({})
    }

    let allTaxes : Taxonomie[] = []

    const updateTaxes = async (taxes: Taxonomie[], id : string, name: string) => {
      if (!taxes || !taxes.length) return
      if (!allTaxes.length) allTaxes = [ ...this.$lodger.taxonomies ]

      taxes.map((tax: Taxonomie) => {
        const $tax = this.$lodger[tax] || this.$lodger[tax.plural]
        if (!$tax) return true

        if (allTaxes && allTaxes.length && allTaxes.indexOf(tax.plural) > -1) {
          allTaxes.splice(allTaxes.indexOf(tax.plural), 1)
        } else {
          // console.error(tax, 'has been handled.')
          return true
        }

        // console.info('DOIN', tax)

        const { subscribers, parents, children } = $tax
        const taxSub = subscribers[subscriberName] ||
          subscribers[mainSubName]

        if (!taxSub) {
          console.error('Invalid subscriber requested', subscriberName, tax)
          return
        }

        let sOrP, op, val

        // deselect selected items of children
        if (taxSub.selectedId) {
          // console.log('deselcting from', tax, taxSub.selectedId)
          taxSub.select(taxSub.selectedId)
        }

        if (taxSub.refsIds) {
          if (parents && parents.length && (parents.indexOf(name) > -1 || parents.indexOf(name.plural) > -1)) {
            const isSingular = parents.indexOf(name) > -1
            sOrP = isSingular ? `${name}Id` : this.form.plural

            op = isSingular ? '$eq' : '$in'
            val = isSingular ? id : [id]

            taxSub.refsIds[sOrP] = val
            // console.log('changed refsIds')
          }
        }

        if (sOrP && op && val) {
          taxSub.criteria.filter = { [sOrP]: { [op]: val } }
          // console.log('updated filter', Object.keys(taxSub.criteria.filter))
        } else {
          if (taxSub.criteria.filter) {
            try {
              delete taxSub.criteria.filter[sOrP]
              // console.log('deleted filter', sOrP)
            } catch (e) {
              console.error('could not delete filter', sOrP, 'on', tax, e)
            }

            if (Object.keys(taxSub.criteria.filter).length === 0) {
              // console.log('completely removed filters')
              taxSub.criteria.filter = null
            }
          }
        }

      })
    }



    reaction(() => sub.selectedId, (id) => {
      allTaxes = [] // has to be reset every time !
      updateTaxes(this.children, id, this.name)
      Object.assign(subState, { selectedId: id })
      // if (this.name === 'utilizator' && this.$lodger) {
      //   state.activeUserId = id
      // }
    })

    // Trigger the modal on activeId change
    reaction(() => sub.activeId, async (id) => {
      if (!id) return
      const activeDoc = await this.collection.findOne(id).exec()
      modal.activeDoc = activeDoc
      Object.assign(modal, { sub })
      Object.assign(subState, { activeId: id })
    })

    reaction(() => ({ ...sub.criteria }), criteria => {
      Object.assign(subState, { criteria: JSON.parse(JSON.stringify(criteria)) })
    })

    if (hooks) {
      // Object.keys(hooks).map(hook => hooks[hook].bind(this))

      // run onEmpty hook
      if (hooks.empty) {
        await sub.updates
        if (!sub.ids.length)
          hooks.empty.call(this)
      }

    }
  }

  /**
   *
   *
   * @param {string} [subscriberName='main']
   * @memberof STaxonomy
   */
  unsubscribe (subscriberName : string = 'main') {
    const subscriber = this.subscribers[subscriberName]

    try {
      if (subscriber.kill)
        subscriber.kill()
    } catch (e) {}

    delete this.subscribers[subscriberName]
  }

  /**
   * Kills all active listeners for a given subscriber name
   *
   * @returns {Promise}
   * @memberof Taxonomy
   */
  protected unsubscribeAll () {
    const { subscribers } = this
    if (!subscribers || !Object.keys(subscribers).length)
      throw new SubscribableTaxonomyError('no subs')
    Object.keys(subscribers).forEach(s => { this.unsubscribe(s) })
  }

  // referencesIds (subName: string) {
  //   const sub = this.subscribers[subName]
  //   if (!sub)
  //     throw new LodgerError('Invalid subscriber requested for refsIds')

  //   const { parents } = this
  //   if (!parents) return

  //   // const x = {}

  //   parents.map(tax => {
  //     const $tax = this.$lodger[tax] || this.$lodger[tax.plural]
  //     if (!$tax) return

  //     const { form: { plural }, subscribers } = $tax
  //     const taxSub = subscribers[subName]

  //     if (taxSub) {
  //       const { selectedId } = taxSub
  //       if (selectedId)
  //         this.refsIds[subName][plural === tax ? plural : `${tax}Id`] = plural === tax ? [ selectedId ] : selectedId
  //     }
  //   })

  //   // return x
  // }

  get hooks () {
    return this.form.taxHooks
  }
}
