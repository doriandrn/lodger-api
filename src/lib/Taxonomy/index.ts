import { RxDocument, RxCollection, RxDatabase, _collectionNamePrimary } from 'rxdb'
import { observable, computed } from 'mobx'

import LodgerConfig from 'lodger.config'
import TaxonomyError from '../Error'
import { LodgerFormCreator, Form } from "../Form"
import notify from '../helpers/notify'

export type TaxonomyCreator<I> = LodgerFormCreator<I>

type LodgerTaxonomyCreatorOptions = {
  multipleSelect ?: boolean,
  timestamps ?: boolean
}

enum Errors {
  noDB = 'Please setup a DB first!'
}

/**
  * Taxonomy item
  *
  * @interface LodgerTaxonomy
  */
interface LodgerTaxonomy<N extends Taxonomie, Interface = {}> {
  readonly collection : RxCollection
  readonly last ?: string

  put (doc: LodgerDocument & Partial<Interface>): Promise<RxDocument<N>> | void
  trash (id: string): Promise<void | null>
}

let db: RxDatabase

/**
 * @class Taxonomy
 * @implements {LodgerTaxonomy}
 *
 * @requires Form
 *
 * @param {Taxonomie} name - name of the form
 * @param {Form} form - the constructed form item
 */
export default class Taxonomy<T extends Taxonomie, Interface = { updatedAt ?: number }>
  implements LodgerTaxonomy<T, Interface> {

  @observable lastItems: string[] = []
  @observable totals: number = 0
  form : Form<Interface>
  $collection ?: RxCollection

  /**
   * Last added item's id
   *
   * @readonly
   * @memberof Taxonomy
   */
  @computed get last () {
    return this.lastItems[0]
  }

  set last (id : string | undefined) {
    if (id) this.lastItems.unshift(id)
    else this.lastItems.shift()
  }

  /**
   * DB handler
   *
   * @static
   * @memberof Taxonomy
   */
  static set db (xdb: RxDatabase) {
    db = xdb
  }

  /**
   * @alias db.destroy
   *
   * @static
   * @memberof Taxonomy
   */
  static async destroy () {
    if (!db) return
    await db.destroy()
  }

  /**
   * Taxonomy name, singular
   *
   * @readonly
   * @memberof Taxonomy
   */
  get name () {
    return this.form.schema.name
  }

  get plural () {
    return this.form.plural
  }

  get _collectionCreator () {
    const { name, methods, statics, } = this._schema
    const { schema } = this.form

    return {
      name,
      schema,
      methods,
      statics
    }
  }

  set collection (collection: RxCollection) {
    const { _schema: { hooks }, options: { timestamps }, $collection } = this

    if ($collection)
      throw new Error('Collection already set for this taxonomy')

    if (timestamps) {
      collection.preSave((data) => {
        data.updatedAt = new Date().getTime()
      }, false)
    }

    // Global hooks
    collection.postInsert((data, doc) => {
      this.totals += 1;
      this.last = doc._id

      doc.atomicUpdate((d) => {
        d.createdAt = new Date().getTime()
        d.updatedAt = data.createdAt
        return d
      })
    }, false)
    collection.postRemove(() => { this.totals -= 1 }, false)

    if (hooks) {
      Object.keys(hooks).forEach(hook => {
        if (['empty'].indexOf(hook) > -1)
          return

        hooks[hook] = hooks[hook](this.$lodger)
        collection[hook](hooks[hook])
      })
    }

    this.form.fieldsIds
      .filter(fieldId => this.form.fields[fieldId].search)
      .forEach(fieldId => collection.searchFields.push(fieldId))

    this.$collection = collection
  }

  get collection () {
    return this.$collection
  }

  /**
   * Creates an instance of Taxonomy.
   *
   * @param {Form<T, Interface>} form
   * @param {RxCollection<Interface>} collection
   * @memberof Taxonomy
   */
  constructor (
    private _schema: LodgerSchema,
    readonly options : LodgerTaxonomyCreatorOptions = { timestamps: true },
  ) {
    const { name, fields, fieldsets, hooks } = _schema
    const { timestamps } = options

    this.form = new Form({ name, fields, fieldsets, hooks }, {
      captureTimestamp: timestamps
    })
  }

  /**
   * Relationship between taxonomies
   *
   * @readonly
   * @memberof Taxonomy
   */
  // get refs () {
  //   const { $lodger, form: { fields, fieldsIds } } = this
  //   if (!$lodger)
  //     return {}

  //   return fieldsIds
  //     .filter(id => fields[id].ref)
  //     .map(id => ({
  //       [id]: fields[id].ref
  //     }))
  // }

  @computed get sortOptions () {
    const { $lodger, form: { fields, schema: { indexes } }, name } = this
    const { i18n } = $lodger

    return Object.assign({},
      ...indexes.map(n => {
        const parsedN = n.replace('.[].', '_')
        const simplifiedN = parsedN.split('_')[0]
        const _18n = i18n.taxonomies[name.plural].fields[simplifiedN] ||
          i18n.sort.fields[simplifiedN] ||
          parsedN

        const field = fields[simplifiedN]
        const { type } = field

        const showOrds = type === 'string' ? ['AZ', 'ZA'] : type === 'number' ? ['09', '90'] : ['asc', 'desc']
        return {
          [ _18n ]: {
            [`${ parsedN }-asc`] : showOrds[0],
            [`${ parsedN }-desc`] :  showOrds[1]
          }
        }
      })
    )
  }

  /**
   * Removes a Document by ID from the collection
   *
   * @param {string} id
   * @returns {RxDocument<T>} removed document
   * @memberof Taxonomy
   */
  async trash (id: string) {
    try {
      await this.collection.findOne(id).remove()
      if (this.last === id) this.last = undefined
    } catch (e) {
      notify({
        type: 'error',
        text: e
      })
    }
  }

  /**
   * Inserts/upserts a new item in DB
   *
   * @param {Object} doc
   * @returns {RxDocument<Taxonomie>} the fresh document
   *
   * @memberof Taxonomy
   */
  async put (
    doc: Partial<Interface> & LodgerDocument
  ) {
    if (!doc || Object.keys(doc).length < 1)
      throw new TaxonomyError('Invalid doc supplied %%', { doc })

    /**
     * If form submitted with an _id, must be an upsert
     */
    const method = doc._id ?
      'upsert' :
      'insert'

    const { name, options } = this

    /**
     * do the insert / upsert and following actions
     */
    try {
      const _doc = await this.collection[method](doc)
      const id = _doc._id

      // this.last = id

      notify({
        type: 'success',
        text: `[${method}] ${name}!${['dev', 'test'].indexOf(process.env.NODE_ENV) > -1 ? `(${id})` : ''}`
      })

      return _doc
    } catch (e) {
      notify({ type: 'error', text: String(e) })
    }
  }

  /**
   * @readonly
   * @memberof Taxonomy
   * Taxonomy default config
   */
  get config () {
    const { taxonomii } = LodgerConfig
    const { defaults } = taxonomii
    return taxonomii[this.name] || defaults
  }
}
