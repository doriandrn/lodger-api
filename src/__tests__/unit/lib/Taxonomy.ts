import Taxonomy from '~/lib/Taxonomy/index'

import { createRxDatabase, addRxPlugin } from 'rxdb'
import sosete from 'fixtures/taxes/sosete'
import testdbsetup from 'fixtures/db/test'

describe('Taxonomy class', () => {
  let $tax: Taxonomy<any>

  beforeAll(async () => {
    addRxPlugin(require('pouchdb-adapter-memory'))
    const db = await createRxDatabase(testdbsetup)
    Taxonomy.db = db
    try {
      $tax = await Taxonomy.init(sosete, { timestamps: true })
    } catch (e) {
      console.log('Taxonomy init failed', e)
    }
  })

  afterAll(async () => {
    await Taxonomy.db.destroy()
  })

  // test('inits with no parameters', async () => {
  //   let x
  //   try {
  //     x = await Taxonomy.init({ name: 'cocosi' })
  //   } catch (e) {
  //     console.error('wtf', e)
  //   }
  //   expect(x).toBeDefined()
  // })

  describe('constructor', () => {
    test('matches snapshot', () => {
      expect($tax).toMatchSnapshot('sosete')
    })

    test('it inits ok for a known tax', () => {
      expect($tax).toBeDefined()
      expect($tax.name).toBe(sosete.name)
    })

    test('form is defined and accesibile', () => {
      expect($tax.form).toBeDefined()
    })
  })

  describe('.put()', () => {
    let soseta, _id

    beforeAll(async () => {
      soseta = await $tax.put({ name: 'verde', lungime: 2 })
      _id = soseta._id
    })

    test('item gets added ok', () => {
      expect(soseta).toBeDefined()
    })

    test('item is assigned an _id', () => {
      expect(soseta._id).toBeDefined()
    })

    test(`getter 'last' is the last added item's id`, async () => {
      const { _id } = soseta

      expect($tax.last).toBe(_id)

      const x = await $tax.put({ name: 'gigi', lungime: 5 })
      expect(x._id).toBe($tax.last)
    })

    test('updates the current item if _id is provided and ok', async () => {
      const name = 'new sos'
      const newSoseta = Object.assign({}, {
        _id,
        name,
        lungime: 3
      })
      const updatedSoseta = await $tax.put(newSoseta)

      expect(_id).toEqual(updatedSoseta._id)
      expect(name).toEqual(updatedSoseta.name)
    })
  })

  describe('.trash()', () => {
    let _id: string

    beforeAll(async () => {
      const x = await $tax.put({ name: 'xx', lungime: 4 })
      _id = x._id
    })

    test('removes ok the item by its id', () => {
      expect(async () => { await $tax.trash(_id) }).not.toThrow()
    })

    test('store last id updates to previous id', () => {
      expect(_id).not.toEqual($tax.last)
    })
  })

  describe('@hooks', () => {
    describe('inited', () => {

    })
  })



  describe('.config', () => {
    const cheiPrincipale = ['limit', 'index', 'sort', 'find']

    test('returneaza default-ul din config - pentru orice taxonomie daca nu e ceruta', () => {
      const { criteriu } = $tax.config
      expect(typeof criteriu).toBe('object')
      const chei = Object.keys(criteriu)
      expect(chei).toEqual(expect.arrayContaining(cheiPrincipale))
    })

    // test('returneaza criteriul cerut pentru taxonomie', () => {

    // })
  })

  // describe('getCriteriu', () => {

  //   test('suprascrie valorile cerute in query', () => {
  //     const limit = 77
  //     const sort = { la: 'lala' }
  //     const criteriu = getCriteriu('asociatie', {
  //       limit,
  //       sort
  //     })
  //     expect(criteriu.limit).toBe(limit)
  //     expect(criteriu.sort).toBe(sort)
  //   })

  //   test('arunca daca e cerut cu string sau altceva', () => {
  //     expect(() => { getCriteriu({}) }).toThrow('taxonomie incorecta')
  //     expect(() => { getCriteriu(23) }).toThrow('taxonomie incorecta')

  //   })

  //   test('returneaza criteriu default pt o taxonomie cunoscuta', () => {
  //     const { limit, index, sort, find } = getCriteriu('asociatii')
  //     expect(limit).toBe(lodgerConfig.taxonomii.asociatii.criteriu.limit)
  //   })
  // })


  // describe('.collection', () => {
  //   describe('positive', () => {
  //     test('makes collection', () => {
  //       const { collection } = __stub1__
  //       expect(collection).toBeDefined()
  //       expect(collection.schema).toBeDefined()
  //     })

  //     test('methods are passed in if existing', () => {
  //       const { collection } = __stub1__
  //       expect(collection.methods).toEqual(stub1.methods)
  //     })

  //     test('statics are passed', () => {
  //       const { collection } = __stub1__
  //       expect(collection.statics).toEqual(stub1.statics)
  //     })
  //   })


  // })

  describe('.children', () => {
    test('can be set', () => {
      $tax.children = [1, 2]
      expect($tax.children).toBeDefined()
    })

    test('throws if not an array', () => {})
  })

  describe('.destroy()', () => {
    beforeAll(async () => {
      await Taxonomy.destroy()
    })

    test('destroys the db', () => {
      expect($tax).toBeDefined()
      expect($tax.db).toBeUndefined()
    })
  })
})
