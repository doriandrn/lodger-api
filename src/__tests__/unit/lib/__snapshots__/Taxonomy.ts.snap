// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Taxonomy class constructor matches snapshot: sosete 1`] = `
Taxonomy {
  "form": Form {
    "$active": false,
    "_onsubmit": Array [
      [Function],
    ],
    "fields": Object {
      "createdAt": Field {
        "default": undefined,
        "index": true,
        "storage": "db",
        "type": "number",
        "value": [Function],
      },
      "lungime": Field {
        "default": undefined,
        "storage": "db",
        "type": "number",
        "value": [Function],
      },
      "name": Field {
        "default": undefined,
        "storage": "db",
        "type": "string",
        "value": [Function],
      },
      "numar": Field {
        "default": undefined,
        "index": true,
        "storage": "db",
        "type": "number",
        "value": [Function],
      },
      "updatedAt": Field {
        "default": undefined,
        "index": true,
        "storage": "db",
        "type": "number",
        "value": [Function],
      },
    },
    "name": "soseta",
    "opts": Object {
      "captureTimestamp": true,
    },
    "plural": "sosetai",
    "schema": Schema {
      "name": "soseta",
      "properties": Object {
        "createdAt": Object {
          "index": true,
          "type": "number",
        },
        "lungime": Object {
          "type": "number",
        },
        "name": Object {
          "type": "string",
        },
        "numar": Object {
          "index": true,
          "type": "number",
        },
        "updatedAt": Object {
          "index": true,
          "type": "number",
        },
      },
      "required": Array [],
      "type": "object",
      "version": 0,
    },
  },
  "lastItems": Array [],
  "options": Object {
    "timestamps": true,
  },
  "refsIds": Array [],
}
`;
